src/nl/hannahsten/texifyidea/inspections/latex/LatexMissingLabelInspection.kt

```diff
@@ -121,32 +152,32 @@ open class LatexMissingLabelInspection : TexifyInspectionBase() {
         }
     }
 
-    private class InsertLabelInEnvironmentFix : LocalQuickFix {
+    private class InsertLabelInEnvironmentFix : LabelQuickFix() {
         override fun getFamilyName() = "Insert label"
 
         override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
             val command = descriptor.psiElement as LatexEnvironment
-            val factory = LatexPsiFactory(project)
-            val labelCommand = LatexPsiFactory.createUniqueLabelCommandFor(factory, command) ?: return
-            if (command.environmentContent == null) {
-                command.addAfter(factory.createEnvironmentContent(), command.beginCommand)
-            }
+            val helper = LatexPsiHelper(project)
+            // Determine label name.
+            val createdLabel = getUniqueLabelName(command.environmentName.formatAsLabel(),
+                    Magic.Environment.labeled[command.environmentName], command.containingFile)
 
-            val environmentContent = command.environmentContent!!
 
-            // in a float environment the label must be inserted after a caption
-            val caption = environmentContent.childrenOfType<LatexCommands>().findLast { c -> c.name == "\\caption" }
-            if (caption != null) {
-                environmentContent.addAfter(labelCommand, caption.parent.parent)
+            if (Magic.Environment.labelAsParameter.contains(command.environmentName)) {
+                helper.addOptionalParameter(command.beginCommand, "label", createdLabel)
             }
             else {
-                environmentContent.add(labelCommand)
+                // in a float environment the label must be inserted after a caption
+                val labelCommand = helper.createLabelCommand(createdLabel)
+                helper.addToContent(command, labelCommand,
+                        command.environmentContent?.childrenOfType<LatexCommands>()
+                                ?.findLast { c -> c.name == "\\caption" })
+
+                // Adjust caret offset
```

Comment:
Interesting, I will have a look at it!

Reply:
It should be fixed now (also for the sections). In the long term it would be nice to offer the possibility to quickly rename the label after insertion (like when you convert an expression to a variable), but I couldn't figure out how to use the templates properly (I think it is done with templates).

Reply:
I still have to test your fix, but would it make sense to have the end result `[label={lst:<caret>}]` instead of `[label={lst:lstlisting2}<caret>]`? Then it is clear that you should provide a better label name than just `lst:lstlisting342`, which I would recommend to do anyway. Same for other environments for which we cannot guess a proper label (like we do for sections).

Reply:
> I still have to test your fix, but would it make sense to have the end result `[label={lst:<caret>}]` instead of `[label={lst:lstlisting2}<caret>]`

That would also be an option. It would basically be a simplistic version of what I described above (IntelliJs rename refactoring when you extract a variable, like in 
![image](https://user-images.githubusercontent.com/7955528/73735227-396a4900-473f-11ea-904f-9c9c5d2861fe.png)). 

The downside of just placing the caret after the prefix instead of generating a default label name and instead of using a proper rename refactoring would be that there is no quick way to just use the default + it might not be entirely obvious to the user that he has to insert a label name now. However, I am also happy to change it to the "just caret after prefix" variant.

Reply:
True, a live template with a default under the cursor would perhaps be nicer, but I don't know how much work that would be to call a live template programmatically and whether a default value is then possible.
If you do want to do something with live templates later, check out #1203 and note that that requires IDEA 2020 (currently on the 2020 branch) to avoid merge conflicts.

Reply:
A default is certainly possible because pressing Ctrl + Alt + V on different expressions produces variables with different default names. I am afraid it is not trivial at least because at least the examples I could find looked rather intricate. If I find some time I will have a look at it again. Thanks for the info regarding #1203 , it looks promising!